### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by tcanabrava**

### CODE REVIEW

File: src/devicemanager/devicemanager.cpp [hunk 1]
```diff
@@ -97,11 +97,10 @@ void DeviceManager::connectLink(LinkConfiguration* linkConf)
 
     // We could use a single Ping instance, but since we are going to support multiple devices
     // this pointer will hold everything for us
-    if(linkConf->deviceType() == PingDeviceType::PING1D) {
-        _primarySensor.reset(new Ping());
-    } else {
-        _primarySensor.reset(new Ping360());
-    }
+     _primarySensor.reset( linkConf->deviceType() == PingDeviceType::PING1D
+        ? static_cast<PingSensor*>(new Ping())
```
Comment (jaxxzer) [inline src/devicemanager/devicemanager.cpp]: my preference is the prior if-else for readability.

File: src/devicemanager/devicemanager.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ void DeviceManager::connectLinkDirectly(AbstractLinkNamespace::LinkType connType
 void DeviceManager::updateAvailableConnections(const QVector<LinkConfiguration>& availableLinkConfigurations)
 {
     // Make all connections unavailable by default
-    for(int i{0}; i < _sensors[Available].size(); i++) {
+    for(int i{0}, end = _sensors[Available].size(); i < end; i++) {
```
Comment (jaxxzer) [inline src/devicemanager/devicemanager.cpp]: squash with 1bc4777906c3a3eb8ba578183ecaf37a1ce85ed5

**Comment by tcanabrava** [comment]
I believe that’s just a matter of getting used to the ternary, it has a
better readability than a if else, The only thing I don’t like there is the
static cast l, but in the end it’s good that it displays the list type.


Em qui, 18 de jul de 2019 às 15:20, Jacob Walser <notifications@github.com>
escreveu:

File: src/flash/flasher.cpp [hunk 3]
```diff
@@ -64,17 +64,17 @@ void Flasher::setVerify(bool verify)
...
-#else
-    static QString absoluteBinPath = QCoreApplication::applicationDirPath();
+        + QStringLiteral("../..")
 #endif
+        + QStringLiteral("/stm32flash")
 #ifdef Q_OS_WIN
-    return absoluteBinPath + "/stm32flash.exe";
-#else
-    return absoluteBinPath + "/stm32flash";
+        + QStringLiteral(".exe");
```
Comment (jaxxzer) [inline src/flash/flasher.cpp]: extra semicolon.

File: src/flash/flasher.cpp [hunk 4]
```diff
@@ -104,7 +104,7 @@ void Flasher::flash()
     static QString cmd = QStringLiteral("\"%0\" -w \"%1\" %2 -g 0x0 -b %3 %4").arg(
                              stm32flashPath(),
                              firmwareFileInfo.absoluteFilePath(),
-                             _verify ? "-v" : "",
+                             _verify ? QStringLiteral("-v") : QString(),
```
Comment (jaxxzer) [inline src/flash/flasher.cpp]: What is the reasoning here? Is this potentially a functional change?

File: src/devicemanager/devicemanager.cpp (hunk 1 above)
Comment (tcanabrava) [reply src/devicemanager/devicemanager.cpp]: actually both versions should die, PingDeviceType is an enum, and thus we need to use an switch.

File: src/flash/flasher.cpp (hunk 4 above)
Comment (tcanabrava) [reply src/flash/flasher.cpp]: Same behavior. QStringLiteral() will transform the strings from utf8 or ascii to utf16 in compile time, and "" will produce a empty string (just with an \0 byte) that will not be appended anywhere because it's empty, while QString() will produce.. nothing. not even a \0 byte. it will not try to convert ascii to utf-16 and thus will execute faster.

File: src/devicemanager/devicemanager.cpp (hunk 1 above)
Comment (jaxxzer) [reply src/devicemanager/devicemanager.cpp]: ok

**Comment by jaxxzer** [comment]
Terenaries are ok, but you placed it inside the function call arguments. My opinion is that that is less readable.

```
call( condition
  ? case1
  : case2
)
```

I wouln't have made the comment if you had done
```
condition
  ? call(case1)
  : call(case2)
```

**Comment by tcanabrava** [comment]
Those are the same thing: a return value will be used in place. While an if
creates an scoped area, and no return value.

Rust fixed that by transforming an it into something that can return values.


Em qui, 18 de jul de 2019 às 15:50, Jacob Walser <notifications@github.com>
escreveu:

File: src/link/link.h [hunk 5]
```diff
@@ -37,7 +37,7 @@ class Link : public QObject
      *
      * @return AbstractLink*
      */
-    AbstractLink* self() { _abstractLink };
+    AbstractLink* self() { return _abstractLink; };
```
Comment (jaxxzer) [inline src/link/link.h]: this belongs to ca39232f92220c6174344a4eaf146a5635c692b6

File: src/link/seriallink.cpp [hunk 6]
```diff
@@ -61,17 +61,18 @@ bool SerialLink::finishConnection()
...
+        currentPortList.append(port.portName());
     }
-    if(oldList != list) {
+
+    if(_availableConnections != currentPortList) {
         emit availableConnectionsChanged();
     }
-    return list;
+
+    return _availableConnections;
```
Comment (jaxxzer) [inline src/link/seriallink.cpp]: The `_availableConnections` value is never updated.

File: src/network/networktool.cpp [hunk 7]
```diff
@@ -12,39 +12,39 @@
...
     //*github.com/user/repo* results in user/repo
     const static QRegularExpression regex(R"(github.com\/([^.]*))");
-    QRegularExpressionMatch regexMatch = regex.match(QStringLiteral(GIT_URL));
+    QRegularExpressionMatch regexMatch = regex.match(userRepoUrl);
     if (!regexMatch.hasMatch() || regexMatch.lastCapturedIndex() < 1) {
-        qCWarning(NETWORKTOOL) <<
-                               "Fail to get github user and repository! "
-                               "It will not be possible to check for updates. "
-                               "Using default value:" << _gitUserRepo;
+        qCWarning(NETWORKTOOL) << "Fail to get github user and repository! "
```
Comment (jaxxzer) [inline src/network/networktool.cpp]: should this one be qstringliteral?

Comment (tcanabrava) [reply src/network/networktool.cpp]: no, qDebug and friends don't need a QString and work witha  char *

File: src/link/seriallink.cpp (hunk 6 above)
Comment (tcanabrava) [reply src/link/seriallink.cpp]: ups. :)

**Comment by jaxxzer** [comment]
@tcanabrava please ping me when the review is addressed and the ci is fixed (run astyle).

File: src/flash/flasher.cpp [hunk 8]
```diff
@@ -12,15 +12,14 @@
 PING_LOGGING_CATEGORY(FLASH, "ping.flash")
 
 Flasher::Flasher(QObject* parent)
-    : QObject(parent)
-{
-    _binRelativePath =
+    : QObject(parent),
+    _binRelativePath(QCoreApplication::applicationDirPath()
```
Comment (patrickelectric) [inline src/flash/flasher.cpp]: use comma in the next line to make it better to read and avoid future unnecessary diffs.
```
: blah
, foo
, bee
```

File: src/network/networktool.cpp [hunk 9]
```diff
@@ -12,39 +12,39 @@
...
-QString NetworkTool::_gitUserRepoFirmware = "bluerobotics/ping-firmware";
-
+namespace {
+    QString gitUserRepo = QStringLiteral("bluerobotics/ping-viewer");
+    QString gitUserRepoFirmware = QStringLiteral("bluerobotics/ping-firmware");
+    const QString gitReleases = QStringLiteral("https://api.github.com/repos/%1/releases");
+    const QString repoContents = QStringLiteral("https://api.github.com/repos/%1/contents/%2");
+    const QString projectTag = QStringLiteral(GIT_TAG);
+    const QString userRepoUrl = QStringLiteral(GIT_URL);
+}
```
Comment (patrickelectric) [inline src/network/networktool.cpp]: Add a space between namespace and class constructor, also the commit does not describes what is happening here and why this namespace is necessary.

File: src/network/networktool.cpp [hunk 10]
```diff
@@ -12,39 +12,39 @@
...
 PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
 
-
-QString NetworkTool::_gitUserRepo = "bluerobotics/ping-viewer";
-QString NetworkTool::_gitUserRepoFirmware = "bluerobotics/ping-firmware";
-
+namespace {
+    QString gitUserRepo = QStringLiteral("bluerobotics/ping-viewer");
+    QString gitUserRepoFirmware = QStringLiteral("bluerobotics/ping-firmware");
+    const QString gitReleases = QStringLiteral("https://api.github.com/repos/%1/releases");
```
Comment (patrickelectric) [inline src/network/networktool.cpp]: maybe: `githubReleasesApiPath`to be more clear, just gitReleases does not imply that the variable is a static path for an API.

File: src/network/networktool.cpp [hunk 11]
```diff
@@ -12,39 +12,39 @@
 
 PING_LOGGING_CATEGORY(NETWORKTOOL, "ping.networktool")
 
-
-QString NetworkTool::_gitUserRepo = "bluerobotics/ping-viewer";
-QString NetworkTool::_gitUserRepoFirmware = "bluerobotics/ping-firmware";
-
+namespace {
+    QString gitUserRepo = QStringLiteral("bluerobotics/ping-viewer");
+    QString gitUserRepoFirmware = QStringLiteral("bluerobotics/ping-firmware");
```
Comment (patrickelectric) [inline src/network/networktool.cpp]: const ?

File: src/link/filelink.cpp [hunk 12]
```diff
@@ -104,9 +104,7 @@ bool FileLink::isOpen()
 bool FileLink::finishConnection()
 {
     // Only close files that are open
-    if(_file.isOpen()) {
-        _file.close();
-    }
+    _file.close();
```
Comment (patrickelectric) [inline src/link/filelink.cpp]: IIRC this will result in a warning message saying that it's closing something that is not open.

File: src/link/link.cpp [hunk 13]
```diff
@@ -37,7 +37,7 @@ Link::Link(LinkType linkType, QString name, QObject* parent)
         _abstractLink.reset(new Ping360SimulationLink());
         break;
     default :
-        qCDebug(PING_PROTOCOL_LINK) << "Link not available!";
+        assert("Link not available!");
```
Comment (patrickelectric) [inline src/link/link.cpp]: We should not assert in something that the user can fix, like choosing a different link or something. `assert` should only be used in places where it's not possible to recover.

File: src/link/link.cpp [hunk 14]
```diff
@@ -19,29 +19,28 @@ Link::Link(LinkType linkType, QString name, QObject* parent)
 {
     switch(linkType) {
     case LinkType::None :
-        _abstractLink.reset(new AbstractLink());
+        _abstractLink = new AbstractLink(this);
```
Comment (patrickelectric) [inline src/link/link.cpp]: The class was created without parent. I don't see the point here. This is also more error prone, E.g: Why are you not deleting `_abstractLink`?

Comment (patrickelectric) [reply src/link/link.cpp]: Also, I'm not aware of such thing as as crashing with Qt and smartpointers, should we stay with smartPointers and remove the setParent ? This appears to be the less error prone option IMO.

File: src/link/seriallink.h [hunk 15]
```diff
@@ -80,4 +80,5 @@ class SerialLink : public AbstractLink
 
 private:
     QSerialPort _port;
+    QStringList _availableConnections;
```
Comment (patrickelectric) [inline src/link/seriallink.h]: alphabetic order, put available before port.

File: src/notification/notificationmodel.cpp [hunk 16]
```diff
@@ -19,11 +19,13 @@ void NotificationModel::add(QVariant text, QVariant icon, QVariant color, QVaria
     const int line = rowCount();
     beginInsertRows(QModelIndex(), line, line);
     insertRows(line, 1);
-    setData(index(line), time, NotificationModel::Time);
-    setData(index(line), text, NotificationModel::Text);
-    setData(index(line), color, NotificationModel::Color);
-    setData(index(line), severity, NotificationModel::Severity);
-    setData(index(line), icon, NotificationModel::Icon);
+    const QModelIndex idx = index(line);
+    /* This should be just *one* call. */
```
Comment (patrickelectric) [inline src/notification/notificationmodel.cpp]: Is this comment a TODO: ?

File: src/sensor/hexvalidator.cpp [hunk 17]
```diff
@@ -31,7 +31,7 @@ bool HexValidator::check(const QByteArray& bytes)
     }
 
     // Check if it has an odd number of elements
-    if(bytes.size()%2 == 0) {
+    if(bytes.size() & 0x1) {
```
Comment (patrickelectric) [inline src/sensor/hexvalidator.cpp]: Don't try to be smarter than the compiler ;)
image

File: src/sensor/parser-ping.cpp [hunk 18]
```diff
@@ -23,12 +23,25 @@ void PingParserExt::parseBuffer(const QByteArray& data)
...
+        case PingParser::ParseState::WAIT_LENGTH_L:
+        case PingParser::ParseState::WAIT_LENGTH_H:
+        case PingParser::ParseState::WAIT_MSG_ID_L:
+        case PingParser::ParseState::WAIT_MSG_ID_H:
+        case PingParser::ParseState::WAIT_SRC_ID:
+        case PingParser::ParseState::WAIT_DST_ID:
+        case PingParser::ParseState::WAIT_PAYLOAD:
+        case PingParser::ParseState::WAIT_CHECKSUM_L:
+        case PingParser::ParseState::WAIT_CHECKSUM_H:
+            return Parser::ParserState::PARSING;
```
Comment (patrickelectric) [inline src/sensor/parser-ping.cpp]: I don't see the point here, the compiler already checks if `==` is a valid enum class and not a int.

File: src/devicemanager/devicemanager.h [hunk 19]
```diff
@@ -30,7 +30,7 @@ class DeviceManager : public QAbstractListModel
      * @return DeviceManager*
      */
     static DeviceManager* self();
-    ~DeviceManager();
+    ~DeviceManager() override;
```
Comment (patrickelectric) [inline src/devicemanager/devicemanager.h]: The commit is not clear. Why is this necessary ? Why override is necessary ? Why are you overriding the desconstrutor. 
Is this overriding the parent deconstruction to avoid any problem related to the singleton nature ?

File: src/flash/flasher.cpp [hunk 20]
```diff
@@ -11,6 +11,9 @@
 
 PING_LOGGING_CATEGORY(FLASH, "ping.flash")
 
+namespace {
+    const QVector<int> validBaudRates = {57600, 115200, 230400};
+}
```
Comment (patrickelectric) [inline src/flash/flasher.cpp]: Please add a space between the namespace and the class constructor.
Why are you moving from privated to namespace ?
Please, be more clear in your commit.

File: src/flash/flasher.h [hunk 21]
```diff
@@ -33,7 +33,7 @@ class Flasher : public QObject
      * @brief Defines flash state
      *
      */
-    enum States {
+    enum class States : char {
```
Comment (patrickelectric) [inline src/flash/flasher.h]: Since this is a numeric representation, use int8_t.

File: src/link/linkconfiguration.h [hunk 22]
```diff
@@ -37,7 +37,7 @@ class LinkConfiguration : public QObject
      *
      */
     enum Error {
-        MissingConfiguration, // This can be used in future for warnings and not real errors
+        MissingConfiguration = -1, // This can be used in future for warnings and not real errors
```
Comment (patrickelectric) [inline src/link/linkconfiguration.h]: good catch.

File: src/link/filelink.cpp [hunk 23]
```diff
@@ -34,9 +34,8 @@ void FileLink::_writeData(const QByteArray& data)
 
     // This save the data as a structure to deal with the timestamp
     if(_openModeFlag == QIODevice::WriteOnly && _file.isWritable()) {
-        QString time =  QTime::fromMSecsSinceStartOfDay(_timer.elapsed()).toString(_timeFormat);
-        Pack pack{time, data};
-        _inout << pack.time << pack.data;
+        _inout << QTime::fromMSecsSinceStartOfDay(_timer.elapsed()).toString(_timeFormat)
```
Comment (patrickelectric) [inline src/link/filelink.cpp]: The usage of the structure here is less error prone, you are creating the structure by hand here, and this does not sync automatically with the Pack definition, resulting in a less readable and more error prone approach.

File: src/sensor/ping.cpp [hunk 24]
```diff
@@ -497,7 +497,7 @@ void Ping::checkNewFirmwareInGitHubPayload(const QJsonDocument& jsonDocument)
     float lastVersionAvailable = 0.0;
 
     auto filesPayload = jsonDocument.array();
-    for(const QJsonValue& filePayload : filesPayload) {
+    for(const QJsonValue filePayload : filesPayload) {
```
Comment (patrickelectric) [inline src/sensor/ping.cpp]: I believe that you want to use QJsonValueRef here since you are moving away from reference.

File: src/devicemanager/devicemanager.cpp [hunk 25]
```diff
@@ -95,10 +97,10 @@ void DeviceManager::connectLink(LinkConfiguration* linkConf)
...
     // We could use a single Ping instance, but since we are going to support multiple devices
     // this pointer will hold everything for us
-    if(linkConf->deviceType() == PingDeviceType::PING1D) {
-        _primarySensor.reset(new Ping());
-    } else {
-        _primarySensor.reset(new Ping360());
+    switch(linkConf->deviceType()) {
+    case PingDeviceType::PING1D : _primarySensor.reset(new Ping()); break;
+    case PingDeviceType::PING360 : _primarySensor.reset(new Ping360()); break;
+    case PingDeviceType::UNKNOWN : assert("Should never happen"); break;
```
[the remark this answers is not in the document]
Comment (tcanabrava) [reply src/devicemanager/devicemanager.cpp]: We can use a Ping1D fallback if you prefer. but if the type goes tehre as unknown it's actually a programming error and aborting at least gives us the backtrace to discover what's wrong.

File: src/flash/flasher.cpp (hunk 20 above)
Comment (tcanabrava) [reply src/flash/flasher.cpp]: I'll rework the commit messages today. anything that's private / static should go into the cpp's unammed namespace, this will reduce the compilation time for the program.

File: src/link/filelink.cpp (hunk 23 above)
Comment (tcanabrava) [reply src/link/filelink.cpp]: ok, I'll ignore this patch.

File: src/sensor/ping.cpp (hunk 24 above)
Comment (tcanabrava) [reply src/sensor/ping.cpp]: Probably. I actually would use nlohmann json for this, seems saner. but right now we are not depending on it and I don't want to add new dependencies.

File: src/flash/flasher.cpp (hunk 8 above)
Comment (tcanabrava) [reply src/flash/flasher.cpp]: this is really bad to read :)

**Comment by patrickelectric** [comment]
@tcanabrava 	please rebase and fix CI when possible.

File: src/link/abstractlink.h [hunk 26]
```diff
@@ -225,7 +225,7 @@ class AbstractLink : public QObject
     const AbstractLink& operator=(const AbstractLink& other);
 
     Q_PROPERTY(qint64 byteSize READ byteSize NOTIFY byteSizeChanged)
-    Q_PROPERTY(LinkConfiguration* configuration READ configuration CONSTANT)
+    Q_PROPERTY(const LinkConfiguration* configuration READ configuration CONSTANT)
```
Comment (patrickelectric) [inline src/link/abstractlink.h]: I believe that this is breaking PingStatus widget:
image

Comment (tcanabrava) [reply src/link/abstractlink.h]: hm.

**Comment by tcanabrava** [comment]
closing the review: too many commits that are unrelated to eachother makes this hard to test. I'll open pull requests that are more coherent.
